### PR TITLE
Added app update functionality

### PR DESCRIPTION
**Pull request by ajmine**

### CODE REVIEW

File: Source/GTInterceptionManager.swift [hunk 1]
```diff
@@ -22,6 +22,45 @@ public class GTInterceptionManager {
...
+            let version: String
+            let promptMessage: Message?
+
+            init(version: String, promptMessage: Message? = nil) {
+                self.version = version
+                self.promptMessage = promptMessage
+            }
+        }
+
+        var active: Bool
```
Comment (MikeSilvis) [inline Source/GTInterceptionManager.swift]: make these `lets`

File: Source/GTInterceptionManager.swift [hunk 2]
```diff
@@ -44,17 +83,108 @@ public class GTInterceptionManager {
...
+    fileprivate func buildUpgradeConfigs(withData data: AnyObject?) {
+        guard let configDicts = data as? [AnyObject] else { return }
+
+        for dict in configDicts {
+            guard let upgradeConfig = upgradeConfig(forDict: dict as? [String: AnyObject]) else { continue }
+            upgradeConfigs.append(upgradeConfig)
+        }
+    }
+
+    fileprivate func upgradeConfig(forDict dict: [String: AnyObject]?) -> UpgradeConfig? {
```
Comment (MikeSilvis) [inline Source/GTInterceptionManager.swift]: nit: `fileprivate func upgradeConfig(dict: [String: AnyObject]?) -> UpgradeConfig? {`

File: Source/GTInterceptionManager.swift [hunk 3]
```diff
@@ -22,6 +22,45 @@ public class GTInterceptionManager {
...
     }
 
+    typealias Message = (title: String, body: String)
+    enum Restriction: Int {
+        case low
+        case medium
+        case high
+    }
+
+    fileprivate struct UpgradeConfig {
```
Comment (MikeSilvis) [inline Source/GTInterceptionManager.swift]: i would put this upgrade stuff into a new class that somehow talks to the interception manager asking for it's data but other then that looks good to me

File: Source/GTInterceptionManager.swift [hunk 4]
```diff
@@ -44,17 +83,108 @@ public class GTInterceptionManager {
...
+        guard !upgradeConfigs.isEmpty else { return (false, nil) }
+
+        func upgradeItem(forConfig upgradeConfig: UpgradeConfig) -> Compatibility? {
+            var violatingVersion: UpgradeConfig.Version?
+
+            if upgradeConfig.minimumAppVersion.version > appVersion {
+                violatingVersion = upgradeConfig.minimumAppVersion
+            } else if upgradeConfig.minimumOSVersion.version > osVersion {
+                violatingVersion = upgradeConfig.minimumOSVersion
+            }
```
Comment (mattbanach) [inline Source/GTInterceptionManager.swift]: it'd be nice if the above was just a nested `func violatingVersion() -> UpgradeConfig.Version?` so you don't need to have a `var violatingVersion`

File: Source/GTInterceptionManager.swift [hunk 5]
```diff
@@ -44,17 +83,108 @@ public class GTInterceptionManager {
...
+    }
+
+    //
+    // MARK : Config Helpers
+    //
+
+    fileprivate func buildUpgradeConfigs(withData data: AnyObject?) {
+        guard let configDicts = data as? [AnyObject] else { return }
+
+        for dict in configDicts {
```
Comment (mattbanach) [inline Source/GTInterceptionManager.swift]: nit: I prefer `forEach { $0.dostuff }`

File: Source/GTInterceptionManager.swift (hunk 4 above)
Comment (MikeSilvis) [reply Source/GTInterceptionManager.swift]: Agreed, I don't ever like changing variables

File: Source/GTUpdateHelper.swift [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+    internal func configureUpdateRequirements(withData data: AnyObject?) {
+        buildUpdateConfigs(withData: data)
+        configureUpdateRequirements()
+    }
+
+    fileprivate func buildUpdateConfigs(withData data: AnyObject?) {
+        guard let configDicts = data as? [[String: AnyObject]] else { return }
+
+        for dict in configDicts {
```
Comment (ajmine) [inline Source/GTUpdateHelper.swift]: @mattbanach ended up leaving this as a `for in` since I couldn't do `continue` in forEach

**Comment by ajmine** [comment]
Added a completion handler for data sync

**Review by MikeSilvis — APPROVED** [review]

**Review by MikeSilvis — APPROVED** [review]

**Pull request merged**